x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -0,0 +1,253 @@
+import org.elasticsearch.index.mapper.SourceToParse;
+import org.elasticsearch.index.mapper.Uid;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardPath;
+import org.elasticsearch.index.snapshots.IndexShardSnapshotStatus;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.recovery.RecoveryState;
+import org.elasticsearch.repositories.FilterRepository;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.repositories.Repository;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.function.Function;
+
+/**
+ * <p>
+ * This is a filter snapshot repository that only snapshots the minimal required information
+ * that is needed to recreate the index. In other words instead of snapshotting the entire shard
+ * with all it's lucene indexed fields, doc values, points etc. it only snapshots the the stored
+ * fields including _source and _routing as well as the live docs in oder to distinguish between
+ * live and deleted docs.
+ * </p>
+ * <p>
+ * The repository can wrap any other repository delegating the source only snapshot to it to and read
+ * from it. For instance a file repository of type <tt>fs</tt> by passing <tt>settings.delegate_type=fs</tt>
+ * at repository creation time.
+ * </p>
+ * The repository supports two distinct restore options:
+ * <ul>
+ *     <li><b>minimal restore</b>: this option re-indexes all documents during restore with an empty mapping. The original mapping is
+ *     stored in the restored indexes <tt>_meta</tt> mapping field. The <tt>minimal</tt> restore must be enabled by setting
+ *     <tt>settings.restore_minimal=true</tt>.</li>
+ *     <li><b>full restore</b>: this option re-indexes all documents during restore with the original mapping. This option is the
+ *     default. This option has a significant operational overhead compared to the minimal option but recreates a fully functional new
+ *     index</li>
+ * </ul>
+ *
+ * Reindex operations are executed in a single thread and can be monitored via indices recovery stats. Every indexed document will be
+ * reported as a translog document.
+ *
+ */
+// TODO: as a followup we should rename translog phase to operation phase in the indices _recovery stats
+public final class SourceOnlySnapshotRepository extends FilterRepository {
+    public static final Setting<String> DELEGATE_TYPE =
+        new Setting<>("delegate_type", "", Function.identity(), Setting.Property.NodeScope);
+
+    public static final Setting<Boolean> RESTORE_MINIMAL = Setting.boolSetting("restore_minimal",
+        false, Setting.Property.NodeScope);
+
+    public static final String SNAPSHOT_DIR_NAME = "_snapshot";
```

Comment:
@bleskes we use tmp dirs for restore and snapshot. I wonder if that is ok or if there are any concerns.  Our shard deletion mechanism should take care of cleaning up.